apps/opencs/view/render/terrainselection.cpp

```diff
@@ -0,0 +1,114 @@
+#include "terrainselection.hpp"
+
+#include <osg/Group>
+
+#include "../../model/world/cellcoordinates.hpp"
+
+#include "worldspacewidget.hpp"
+
+namespace
+{
+    const int cellSize {ESM::Land::REAL_SIZE};
+    const int landSize {ESM::Land::LAND_SIZE};
+    const int landTextureSize {ESM::Land::LAND_TEXTURE_SIZE};
+}
+
+CSVRender::TerrainSelection::TerrainSelection(const CSMWorld::CellCoordinates& coords, const ESM::Land& esmLand, osg::Group* parentNode)
+    :mCoords {coords}, mEsmLand {esmLand}, mParentNode {parentNode}
+{
+    mBaseNode = new osg::PositionAttitudeTransform{};
+    mBaseNode->setPosition(osg::Vec3d(coords.getX() * cellSize, coords.getY() * cellSize, 10.0));
+    activate();
+}
+
+CSVRender::TerrainSelection::~TerrainSelection()
+{
+    deactivate();
+}
+
+void CSVRender::TerrainSelection::select(const WorldspaceHitResult& hit)
+{
+    deselect();
+
+    if (isInCell(hit)) {
+        addToSelection(hit.worldPos);
+    }
+    update();
+}
+
+void CSVRender::TerrainSelection::toggleSelect(const WorldspaceHitResult& hit)
+{
+    if (isInCell(hit)) {
+        toggleSelection(hit.worldPos);
+        update();
+    }
+}
+
+void CSVRender::TerrainSelection::activate()
+{
+    mParentNode->addChild(mBaseNode);
```

Comment:
Is there any way activate() could potentially be called twice? In that case we need to add a check to make sure mBaseNode won't be added to the scene twice. Even if that doesn't happen, we should probably still add the check just to be sure.

Reply:
From <osg/Group> header:
> If node is not NULL and is not contained in Group then increment its reference count, add it to the child list and dirty the bounding sphere to force it to recompute on next getBound() and return true for success. Otherwise return false.

It seems that if a node is already contained within a Group, adding it again is just a no-op. Does this mean that no check is needed?

Reply:
Well, that documentation appears to be wrong. If you look at the [source](https://github.com/openscenegraph/OpenSceneGraph/blob/master/src/osg/Group.cpp#L77), it will only check for duplicates if ENSURE_CHILD_IS_UNIQUE is defined, which is never defined anywhere.